File: pkg/java/src/main/java/dev/openfga/language/OpenFgaDslListener.java
package dev.openfga.language;

import dev.openfga.language.antlr.OpenFGAParser;
import dev.openfga.language.antlr.OpenFGAParserBaseListener;
import dev.openfga.sdk.api.model.*;

import java.util.*;

public class OpenFgaDslListener extends OpenFGAParserBaseListener {
    private static final String RELATION_DEFINITION_OPERATOR_OR = "or";
    private static final String RELATION_DEFINITION_OPERATOR_AND = "and";
    private static final String RELATION_DEFINITION_OPERATOR_BUT_NOT = "but not";

    private final AuthorizationModel authorizationModel = new AuthorizationModel();
    private final OpenFGAParser parser;
    private TypeDefinition currentTypeDef = null;
    private Relation currentRelation = null;
    private Condition currentCondition = null;

    private Deque<StackRelation> rewriteStack = null;

    public OpenFgaDslListener(OpenFGAParser parser) {
        this.parser = parser;
    }

    public AuthorizationModel getAuthorizationModel() {
        return authorizationModel;
    }

    private Userset parseExpression(List<Userset> rewrites, String operator) {

        if (rewrites.isEmpty()) {
            return null;
        }
        if (rewrites.size() == 1) {
            return rewrites.get(0);
        }
        Userset relationDef = null;
        switch (operator) {
            case RELATION_DEFINITION_OPERATOR_OR:
                relationDef = new Userset().union(new Usersets().child(rewrites));
                break;
            case RELATION_DEFINITION_OPERATOR_AND:
                relationDef = new Userset().intersection(new Usersets().child(rewrites));
                break;
            case RELATION_DEFINITION_OPERATOR_BUT_NOT:
                relationDef = new Userset().difference(new Difference().base(rewrites.get(0)).subtract(rewrites.get(1)));
                break;
        }
        return relationDef;
    }

    @Override
    public void enterMain(OpenFGAParser.MainContext ctx) {
        authorizationModel.setConditions(new HashMap<>());
    }

    @Override
    public void exitModelHeader(OpenFGAParser.ModelHeaderContext ctx) {
        if (ctx.SCHEMA_VERSION() != null) {
            authorizationModel.setSchemaVersion(ctx.SCHEMA_VERSION().getText());
        }
    }

    @Override
    public void enterTypeDefs(OpenFGAParser.TypeDefsContext ctx) {
        this.authorizationModel.setTypeDefinitions(new ArrayList<>());
    }

    @Override
    public void enterTypeDef(OpenFGAParser.TypeDefContext ctx) {
        if (ctx.typeName == null) {
            return;
        }

        currentTypeDef = new TypeDefinition()
                .type(ctx.typeName.getText())
                .relations(new HashMap<>())
                .metadata(new Metadata().relations(new HashMap<>()));

    }

    @Override
    public void enterConditions(OpenFGAParser.ConditionsContext ctx) {
        authorizationModel.setConditions(new HashMap<>());
    }

    @Override
    public void enterCondition(OpenFGAParser.ConditionContext ctx) {
        if (ctx.conditionName() == null) {
            return;
        }

        var conditionName = ctx.conditionName().getText();
        if (authorizationModel.getConditions().containsKey(conditionName)) {
            var message = String.format("condition '%s' is already defined in the model", conditionName);
            parser.notifyErrorListeners(ctx.conditionName().start, message, null);
        }

        currentCondition = new Condition()
                .name(conditionName)
                .expression("")
                .parameters(new HashMap<>());
    }

    @Override
    public void exitConditionParameter(OpenFGAParser.ConditionParameterContext ctx) {
        if (ctx.parameterName() == null || ctx.parameterType() == null) {
            return;
        }

        var parameterName = ctx.parameterName().getText();
        if (currentCondition.getParameters().containsKey(parameterName)) {
            var message = String.format("parameter '%s' is already defined in the condition '%s'", parameterName, currentCondition.getName());
            parser.notifyErrorListeners(ctx.parameterName().start, message, null);
        }

        var paramContainer = ctx.parameterType().CONDITION_PARAM_CONTAINER();
        var conditionParamTypeRef = new PartialConditionParamTypeRef();
        var typeName = ctx.parameterType().getText();
        if (paramContainer != null) {
            typeName = paramContainer.getText();
            conditionParamTypeRef.setTypeName(parseTypeName(paramContainer.getText()));
            if (ctx.parameterType().CONDITION_PARAM_TYPE() != null) {
                var genericTypeName = parseTypeName(ctx.parameterType().CONDITION_PARAM_TYPE().getText());
                if (genericTypeName != TypeName.UNKNOWN_DEFAULT_OPEN_API) {
                    conditionParamTypeRef.setGenericTypes(
                            new ArrayList<>() {{
                                add(new ConditionParamTypeRef().typeName(genericTypeName));
                            }}
                    );
                }

            }
        }
        conditionParamTypeRef.setTypeName(parseTypeName(typeName));

        currentCondition.getParameters().put(parameterName, conditionParamTypeRef.asConditionParamTypeRef());
    }

    private TypeName parseTypeName(String typeName) {
        return TypeName.fromValue("TYPE_NAME_" + typeName.toUpperCase());
    }
    @Override
    public void exitConditionExpression(OpenFGAParser.ConditionExpressionContext ctx) {
        currentCondition.setExpression(ctx.getText().trim());
    }

    @Override
    public void exitCondition(OpenFGAParser.ConditionContext ctx) {
        if (currentCondition != null) {
            authorizationModel.getConditions().put(currentCondition.getName(), currentCondition);
            currentCondition = null;
        }
    }

    @Override
    public void exitTypeDef(OpenFGAParser.TypeDefContext ctx) {
        if (currentTypeDef == null) {
            return;
        }

        if (currentTypeDef.getMetadata() != null
                && currentTypeDef.getMetadata().getRelations() != null
                && currentTypeDef.getMetadata().getRelations().isEmpty()) {
            currentTypeDef.setMetadata(null);
        }

        var typeDefinitions = authorizationModel.getTypeDefinitions();
        if (typeDefinitions != null) {
            typeDefinitions.add(currentTypeDef);
        }

        currentTypeDef = null;
    }

    @Override
    public void enterRelationDeclaration(OpenFGAParser.RelationDeclarationContext ctx) {
        currentRelation = new Relation(
                null,
                new ArrayList<>(),
                null,
                new RelationMetadata().directlyRelatedUserTypes(new ArrayList<>()));
        rewriteStack = new ArrayDeque<>();
    }

    @Override
    public void exitRelationDeclaration(OpenFGAParser.RelationDeclarationContext ctx) {
        if (ctx.relationName() == null) {
            return;
        }

        var relationName = ctx.relationName().getText();

        var relationDef = parseExpression(currentRelation.getRewrites(), currentRelation.getOperator());
        if(relationDef != null) {
            if (this.currentTypeDef.getRelations().get(relationName) != null) {
                var message = String.format("'%s' is already defined in '%s'", relationName, currentTypeDef.getType());
                parser.notifyErrorListeners(ctx.relationName().start, message, null);
            }

            currentTypeDef.getRelations().put(relationName, relationDef);
            var directlyRelatedUserTypes = currentRelation.getTypeInfo().getDirectlyRelatedUserTypes();
            currentTypeDef.getMetadata().getRelations().put(relationName, new RelationMetadata().directlyRelatedUserTypes(directlyRelatedUserTypes));
        }

        currentRelation = null;
    }

    @Override
    public void enterRelationDefDirectAssignment(OpenFGAParser.RelationDefDirectAssignmentContext ctx) {
        currentRelation.setTypeInfo(new RelationMetadata().directlyRelatedUserTypes(new ArrayList<>()));
    }

    @Override
    public void exitRelationDefDirectAssignment(OpenFGAParser.RelationDefDirectAssignmentContext ctx) {
        var partialRewrite = new Userset()._this(new HashMap<>());
        currentRelation.getRewrites().add(partialRewrite);
    }

    @Override
    public void exitRelationDefTypeRestriction(OpenFGAParser.RelationDefTypeRestrictionContext ctx) {

        var baseRestriction = ctx.relationDefTypeRestrictionBase();
        if (baseRestriction == null) {
            return;
        }

        var _type = baseRestriction.relationDefTypeRestrictionType;
        var usersetRestriction = baseRestriction.relationDefTypeRestrictionRelation;
        var wildcardRestriction = baseRestriction.relationDefTypeRestrictionWildcard;
        var conditionName = ctx.conditionName();

        var relationRef = new PartialRelationReference();
        if (_type != null) {
            relationRef.setType(_type.getText());
        }

        if(conditionName != null) {
            relationRef.setCondition(conditionName.getText());
        }

        if(usersetRestriction != null) {
            relationRef.setRelation(usersetRestriction.getText());
        }

        if(wildcardRestriction != null) {
            relationRef.setWildcard(new HashMap<>());
        }

        currentRelation.getTypeInfo().getDirectlyRelatedUserTypes().add(relationRef.asRelationReference());
    }

    @Override
    public void exitRelationDefRewrite(OpenFGAParser.RelationDefRewriteContext ctx) {
        var computedUserset = new ObjectRelation().relation(ctx.rewriteComputedusersetName.getText());

        var partialRewrite = ctx.rewriteTuplesetName == null
                ? new Userset().computedUserset(computedUserset)
                : new Userset().tupleToUserset(new TupleToUserset()
                .computedUserset(computedUserset)
                .tupleset(new ObjectRelation().relation(ctx.rewriteTuplesetName.getText()))
        );

        currentRelation.getRewrites().add(partialRewrite);
    }

    @Override
    public void exitRelationRecurse(OpenFGAParser.RelationRecurseContext ctx) {
        if (currentRelation == null) {
            return;
        }

        var relationDef = parseExpression(currentRelation.getRewrites(), currentRelation.getOperator());

        if (relationDef != null) {
            currentRelation.setRewrites(new ArrayList<>() {{
                add(relationDef);
            }});
        }
    }

    @Override
    public void enterRelationRecurseNoDirect(OpenFGAParser.RelationRecurseNoDirectContext ctx) {
        if (rewriteStack != null) {
            rewriteStack.add(new StackRelation(currentRelation.getRewrites(), currentRelation.getOperator()));
        }

        currentRelation.setRewrites(new ArrayList<>());
    }

    @Override
    public void exitRelationRecurseNoDirect(OpenFGAParser.RelationRecurseNoDirectContext ctx) {
        if (currentRelation == null) {
            return;
        }

        var popped = rewriteStack.removeLast();

        var relationDef = parseExpression(currentRelation.getRewrites(), currentRelation.getOperator());
        if (relationDef != null) {
            currentRelation.setOperator(popped.getOperator());
            currentRelation.setRewrites(new ArrayList<>(popped.getRewrites()) {{
                add(relationDef);
            }});
        }
    }

    @Override
    public void enterRelationDefPartials(OpenFGAParser.RelationDefPartialsContext ctx) {
        if (!ctx.OR().isEmpty()) {
            currentRelation.setOperator(RELATION_DEFINITION_OPERATOR_OR);
        } else if (!ctx.AND().isEmpty()) {
            currentRelation.setOperator(RELATION_DEFINITION_OPERATOR_AND);
        } else if (ctx.BUT_NOT() != null) {
            currentRelation.setOperator(RELATION_DEFINITION_OPERATOR_BUT_NOT);
        }
    }
}

File: pkg/csharp/src/OpenFga.Language/OpenFgaDslListener.cs
using OpenFga.Sdk.Model;

namespace OpenFga.Language;

/// <summary>
/// OpenFGA DSL Listener that uses OpenFGA SDK model classes
/// This is the main version that works with the actual SDK
/// </summary>
public class OpenFgaDslListener(OpenFGAParser parser)
    : OpenFGAParserBaseListener {

    private const string RelationDefinitionOperatorOr = "or";
    private const string RelationDefinitionOperatorAnd = "and";
    private const string RelationDefinitionOperatorButNot = "but not";

    private readonly AuthorizationModel _authorizationModel = new();
    private TypeDefinition? _currentTypeDef;
    private Relation? _currentRelation;
    private Condition? _currentCondition;
    private bool _isModularModel;
    private readonly Dictionary<string, TypeDefinition> _typeDefExtensions = new();

    private Stack<StackRelation>? _rewriteStack;

    public AuthorizationModel GetAuthorizationModel() {
        return _authorizationModel;
    }

    private Userset? ParseExpression(List<Userset> rewrites, string? @operator) {
        if (rewrites == null || !rewrites.Any()) {
            return null;
        }

        if (rewrites.Count == 1) {
            return rewrites[0];
        }

        Userset? relationDef = null;
        switch (@operator) {
            case RelationDefinitionOperatorOr:
                relationDef = new Userset {
                    Union = new Usersets {
                        Child = rewrites
                    }
                };
                break;
            case RelationDefinitionOperatorAnd:
                relationDef = new Userset {
                    Intersection = new Usersets {
                        Child = rewrites
                    }
                };
                break;
            case RelationDefinitionOperatorButNot:
                relationDef = new Userset {
                    Difference = new Difference {
                        Base = rewrites[0],
                        Subtract = rewrites[1]
                    }
                };
                break;
        }

        return relationDef;
    }

    public override void EnterMain(OpenFGAParser.MainContext context) {
        _authorizationModel.Conditions = new Dictionary<string, Condition>();
        base.EnterMain(context);
    }

    public override void ExitMain(OpenFGAParser.MainContext context) {
        // TO MAKE TEST PASS: If there are no conditions, set the Conditions to null
        if ((_authorizationModel.Conditions?.Count ?? 0) == 0) {
            _authorizationModel.Conditions = null;
        }
        base.ExitMain(context);
    }

    public override void ExitModelHeader(OpenFGAParser.ModelHeaderContext context) {
        if (context.SCHEMA_VERSION() != null) {
            _authorizationModel.SchemaVersion = context.SCHEMA_VERSION().GetText();
        }

        base.ExitModelHeader(context);
    }

    public override void ExitModuleHeader(OpenFGAParser.ModuleHeaderContext context) {
        this._isModularModel = true;
        base.ExitModuleHeader(context);
    }

    public override void EnterTypeDefs(OpenFGAParser.TypeDefsContext context) {
        this._authorizationModel.TypeDefinitions = new List<TypeDefinition>();
        base.EnterTypeDefs(context);
    }

    public override void ExitTypeDefs(OpenFGAParser.TypeDefsContext context) {
        // TO MAKE TEST PASS: If there are no type definitions, set the TypeDefinitions to null
        if (this._authorizationModel.TypeDefinitions.Count == 0) {
            this._authorizationModel.TypeDefinitions = null!;
        }
        base.ExitTypeDefs(context);
    }

    public override void EnterTypeDef(OpenFGAParser.TypeDefContext context) {
        if (context.typeName == null) {
            return;
        }

        if (context.EXTEND() != null && !this._isModularModel) {
            parser.NotifyErrorListeners(context.typeName.Start, "extend can only be used in a modular model", null);
        }

        _currentTypeDef = new TypeDefinition {
            Type = context.typeName.GetText(),
            Relations = new Dictionary<string, Userset>(),
            Metadata = new Metadata {
                Relations = new Dictionary<string, RelationMetadata>()
            }
        };

        base.EnterTypeDef(context);
    }

    public override void EnterConditions(OpenFGAParser.ConditionsContext context) {
        _authorizationModel.Conditions = new Dictionary<string, Condition>();
        base.EnterConditions(context);
    }

    public override void EnterCondition(OpenFGAParser.ConditionContext context) {
        if (context.conditionName() == null) {
            return;
        }

        var conditionName = context.conditionName().GetText();
        if (_authorizationModel.Conditions?.ContainsKey(conditionName) ?? false) {
            var message = $"condition '{conditionName}' is already defined in the model";
            parser.NotifyErrorListeners(context.conditionName().Start, message, null);
        }

        _currentCondition = new Condition {
            Name = conditionName,
            Expression = "",
            Parameters = new Dictionary<string, ConditionParamTypeRef>()
        };

        base.EnterCondition(context);
    }

    public override void ExitConditionParameter(OpenFGAParser.ConditionParameterContext context) {
        if (context.parameterName() == null || context.parameterType() == null) {
            return;
        }

        var parameterName = context.parameterName().GetText();
        if (_currentCondition!.Parameters!.ContainsKey(parameterName)) {
            var message = string.Format(
                "parameter '{0}' is already defined in the condition '{1}'",
                parameterName, _currentCondition.Name);
            parser.NotifyErrorListeners(context.parameterName().Start, message, null);
        }

        var paramContainer = context.parameterType().CONDITION_PARAM_CONTAINER();
        var conditionParamTypeRef = new PartialConditionParamTypeRef();
        var typeName = context.parameterType().GetText();
        if (paramContainer != null) {
            typeName = paramContainer.GetText();
            conditionParamTypeRef.TypeName = ParseTypeName(paramContainer.GetText());
            if (context.parameterType().CONDITION_PARAM_TYPE() != null) {
                var genericTypeName =
                    ParseTypeName(context.parameterType().CONDITION_PARAM_TYPE().GetText());
                if (genericTypeName != (TypeName)13) {
                    conditionParamTypeRef.GenericTypes =
                    [
                        new ConditionParamTypeRef()
                        {
                            TypeName = genericTypeName
                        }
                    ];
                }
            }
        }

        conditionParamTypeRef.TypeName = ParseTypeName(typeName);

        _currentCondition.Parameters[parameterName] = conditionParamTypeRef.AsConditionParamTypeRef();

        base.ExitConditionParameter(context);
    }

    private TypeName ParseTypeName(string typeName) {
        return Enum.Parse<TypeName>("TYPENAME" + typeName.ToUpper(), true);
    }

    public override void ExitConditionExpression(OpenFGAParser.ConditionExpressionContext context) {
        if (_currentCondition != null) {
            _currentCondition.Expression = context.GetText().Trim();
        }

        base.ExitConditionExpression(context);
    }

    public override void ExitCondition(OpenFGAParser.ConditionContext context) {
        if (_currentCondition != null) {
            _authorizationModel.Conditions![_currentCondition.Name] = _currentCondition;
            _currentCondition = null;
        }

        base.ExitCondition(context);
    }

    public override void ExitTypeDef(OpenFGAParser.TypeDefContext context) {
        if (_currentTypeDef == null) {
            return;
        }

        if (_currentTypeDef.Metadata is { Relations: not null }
            && !_currentTypeDef.Metadata.Relations.Any()) {
            _currentTypeDef.Metadata = null;
        }

        var typeDefinitions = _authorizationModel.TypeDefinitions;
        if (typeDefinitions != null) {
            typeDefinitions.Add(_currentTypeDef);
        }

        if (context.EXTEND() != null && this._isModularModel) {
            if (_typeDefExtensions.ContainsKey(_currentTypeDef.Type)) {
                parser.NotifyErrorListeners(
                    context.typeName.Start,
                    string.Format("'{0}' is already extended in file.", _currentTypeDef.Type),
                    null);
            }
            else {
                _typeDefExtensions[_currentTypeDef.Type] = _currentTypeDef;
            }
        }

        _currentTypeDef = null;
        base.ExitTypeDef(context);
    }

    public override void EnterRelationDeclaration(OpenFGAParser.RelationDeclarationContext context) {
        _currentRelation = new Relation(
            null,
            new List<Userset>(),
            null,
            new RelationMetadata {
                DirectlyRelatedUserTypes = new List<RelationReference>()
            });
        _rewriteStack = new Stack<StackRelation>();

        base.EnterRelationDeclaration(context);
    }

    public override void ExitRelationDeclaration(OpenFGAParser.RelationDeclarationContext context) {
        if (context.relationName() == null || _currentRelation == null || _currentTypeDef == null) {
            return;
        }

        var relationName = context.relationName().GetText();

        var relationDef = ParseExpression(_currentRelation.Rewrites, _currentRelation.Operator);
        if (relationDef != null) {
            if (_currentTypeDef.Relations!.ContainsKey(relationName)) {
                var message = $"'{relationName}' is already defined in '{_currentTypeDef.Type}'";
                parser.NotifyErrorListeners(context.relationName().Start, message, null);
            }

            _currentTypeDef.Relations[relationName] = relationDef;
            var directlyRelatedUserTypes = _currentRelation.TypeInfo.DirectlyRelatedUserTypes;
            if (_currentTypeDef.Metadata?.Relations != null) {
                _currentTypeDef.Metadata.Relations[relationName] = new RelationMetadata {
                    DirectlyRelatedUserTypes = directlyRelatedUserTypes
                };
            }
        }

        _currentRelation = null;
        base.ExitRelationDeclaration(context);
    }

    public override void EnterRelationDefDirectAssignment(OpenFGAParser.RelationDefDirectAssignmentContext context) {
        if (_currentRelation != null) {
            _currentRelation.TypeInfo = new RelationMetadata {
                DirectlyRelatedUserTypes = new List<RelationReference>()
            };
        }

        base.EnterRelationDefDirectAssignment(context);
    }

    public override void ExitRelationDefDirectAssignment(OpenFGAParser.RelationDefDirectAssignmentContext context) {
        if (_currentRelation != null) {
            var partialRewrite = new Userset {
                This = new Dictionary<string, object>()
            };
            _currentRelation.Rewrites.Add(partialRewrite);
        }

        base.ExitRelationDefDirectAssignment(context);
    }

    public override void ExitRelationDefTypeRestriction(OpenFGAParser.RelationDefTypeRestrictionContext context) {
        if (_currentRelation == null) {
            return;
        }

        var baseRestriction = context.relationDefTypeRestrictionBase();
        if (baseRestriction == null) {
            return;
        }

        var t = baseRestriction.relationDefTypeRestrictionType;
        var usersetRestriction = baseRestriction.relationDefTypeRestrictionRelation;
        var wildcardRestriction = baseRestriction.relationDefTypeRestrictionWildcard;
        var conditionName = context.conditionName();

        var relationRef = new PartialRelationReference();
        if (t != null) {
            relationRef.Type = t.GetText();
        }

        if (conditionName != null) {
            relationRef.Condition = conditionName.GetText();
        }

        if (usersetRestriction != null) {
            relationRef.Relation = usersetRestriction.GetText();
        }

        if (wildcardRestriction != null) {
            relationRef.Wildcard = new Dictionary<string, object>();
        }

        _currentRelation.TypeInfo.DirectlyRelatedUserTypes!.Add(relationRef.AsRelationReference());
        base.ExitRelationDefTypeRestriction(context);
    }

    public override void ExitRelationDefRewrite(OpenFGAParser.RelationDefRewriteContext context) {
        if (_currentRelation == null) {
            return;
        }

        var computedUserset = new ObjectRelation {
            Relation = context.rewriteComputedusersetName.GetText()
        };

        Userset partialRewrite;
        if (context.rewriteTuplesetName == null) {
            partialRewrite = new Userset {
                ComputedUserset = computedUserset
            };
        }
        else {
            partialRewrite = new Userset {
                TupleToUserset = new TupleToUserset {
                    ComputedUserset = computedUserset,
                    Tupleset = new ObjectRelation {
                        Relation = context.rewriteTuplesetName.GetText()
                    }
                }
            };
        }

        _currentRelation.Rewrites.Add(partialRewrite);
        base.ExitRelationDefRewrite(context);
    }

    public override void ExitRelationRecurse(OpenFGAParser.RelationRecurseContext context) {
        if (_currentRelation == null) {
            return;
        }

        var relationDef = ParseExpression(_currentRelation.Rewrites, _currentRelation.Operator);

        if (relationDef != null) {
            _currentRelation.Rewrites = new List<Userset> { relationDef };
        }

        base.ExitRelationRecurse(context);
    }

    public override void EnterRelationRecurseNoDirect(OpenFGAParser.RelationRecurseNoDirectContext context) {
        if (_rewriteStack != null && _currentRelation != null) {
            _rewriteStack.Push(new StackRelation(_currentRelation.Rewrites, _currentRelation.Operator));
        }

        if (_currentRelation != null) {
            _currentRelation.Rewrites = new List<Userset>();
        }

        base.EnterRelationRecurseNoDirect(context);
    }

    public override void ExitRelationRecurseNoDirect(OpenFGAParser.RelationRecurseNoDirectContext context) {
        if (_currentRelation == null || _rewriteStack == null) {
            return;
        }

        var popped = _rewriteStack.Pop();

        var relationDef = ParseExpression(_currentRelation.Rewrites, _currentRelation.Operator);
        if (relationDef != null) {
            _currentRelation.Operator = popped.Operator;
            _currentRelation.Rewrites = new List<Userset>(popped.Rewrites) { relationDef };
        }

        base.ExitRelationRecurseNoDirect(context);
    }

    public override void EnterRelationDefPartials(OpenFGAParser.RelationDefPartialsContext context) {
        if (_currentRelation != null) {
            if (context.OR() != null && context.OR().Length > 0) {
                _currentRelation.Operator = RelationDefinitionOperatorOr;
            }
            else if (context.AND() != null && context.AND().Length > 0) {
                _currentRelation.Operator = RelationDefinitionOperatorAnd;
            }
            else if (context.BUT_NOT() != null) {
                _currentRelation.Operator = RelationDefinitionOperatorButNot;
            }
        }

        base.EnterRelationDefPartials(context);
    }
}